src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java

```diff
@@ -656,69 +661,220 @@ private static void filterAndAddKeys(String metricName, SetMultimap<String, Stri
+                metricName,
+                startTime,
+                endTime,
+                filterTags,
+                limit);
+        return rowKeys;
+    }
 
-            //  If any part of the query window falls before the new split index
-            //  start time, use the old split index:
-            if (m_cassandraConfiguration.isUseNewSplitIndexRead() && startTime > m_cassandraConfiguration.getNewSplitIndexStartTimeMs()) {
-                logger.info("Using new split index for query starting at {}.", startTime);
-                bs = m_psQueryRowKeySplitIndex2.bind();
+    private List<DataPointsRowKey> getMatchingRowKeysFromSplitIndex(String metricName,
+                                                                    String useSplitField,
+                                                                    List<String> useSplit,
+                                                                    long startTime,
+                                                                    long endTime,
+                                                                    SetMultimap<String, String> filterTags,
+                                                                    int limit) {
+        final List<DataPointsRowKey> rowKeys = new LinkedList<>();
+        if (m_cassandraConfiguration.isUseNewSplitIndexRead() && endTime >= m_cassandraConfiguration.getNewSplitIndexStartTimeMs()) {
+            // Allowed cases for this branch (* - a point defined by m_cassandraConfiguration.getNewSplitIndexStartTimeMs())
+            //  --|-------*------|--
+            // startTime     endTime
+            //
+            //  --|--------------*--
+            // startTime     endTime
+            //
+            //  *--------|--------------|--
+            //        startTime     endTime
+            //
+            //  --*--------------|--
+            //  startTime     endTime
+
+            if (startTime < m_cassandraConfiguration.getNewSplitIndexStartTimeMs()) {
+                // cases
+                //  --|-------*------|--
+                // startTime     endTime
+                //  --|--------------*--
+                // startTime     endTime
+
+                // in case m_cassandraConfiguration.getNewSplitIndexStartTimeMs() == endTime
+                // with first round it will fetch data from new table for a single time point == 'endTime'
+                // the second round will fetch all the data until 'endTime' (exclusively)
+                collectFromRowKeySplitIndex(rowKeys,
+                        m_psQueryRowKeySplitIndex2,
+                        metricName,
+                        useSplitField,
+                        useSplit,
+                        m_cassandraConfiguration.getNewSplitIndexStartTimeMs(),
+                        endTime,
+                        filterTags,
+                        limit);
+
+                limit -= rowKeys.size();
+                if (limit > 0) {
```

Comment:
Can't this result in a scenario where we don't show old, historical data because there is no recent data?

Reply:
if there is no recent data, then `limit` will be decreased by 0 and old data will be fetched